### ISSUE TITLE
pyperfmon querying counters

### ISSUE THREAD

**Issue by rpratt-eng1** [damies13/pyperfmon#1]
Hi,
I wrote a test script using pyperfmon to monitor performance of various disk counters on a remote machine. It is working, but I wanted a loop to be able poll at a certain interval a measure the performance at that current time. However, in running the script, once it does the initial connect to the remote server, it keeps reporting the same number over and over. It appears that the data is cached and it isn't actually querying the counter for new data. I've noticed, it happens locally as well.  I don't think it's a problem inherent with pyperfmon, might be and might not be, but that having been said, have you run into this issue before and what did you do to resolve the issue? 

ex: 
rt = pm.getCounter(r"PhysicalDisk\_Total\Disk Read Bytes/sec","192.168.1.99")
('PhysicalDisk\\_Total\\Disk Bytes/sec', '589363867')
time.sleep(5)
('PhysicalDisk\\_Total\\Disk Bytes/sec', '589363867')
time.sleep(5)
('PhysicalDisk\\_Total\\Disk Bytes/sec', '589363867')
time.sleep(5)
('PhysicalDisk\\_Total\\Disk Bytes/sec', '589363867')

monitoring perfmon on the server shows this result does not remain constant. 

Thanks!

**Comment by damies13** [comment]
Thanks for the issue,

That's interesting, all pyperfmon does is query for the current value, so I would expect it to change if it's actually changing on the machine being monitored. I'll try to reproduce this.

**Comment by k5md** [comment]
This happens because of how pyperfmon uses wmi objects (and how wmi works). You can read about this problem [here](https://devblogs.microsoft.com/scripting/why-does-my-performance-monitoring-script-keep-returning-the-same-incorrect-values/), or compare two scripts. This one will work correctly:
```
import wmi
from time import sleep
c = wmi.WMI(".")
while True:
    mem = c.Win32_PerfRawData_PerfOS_Memory()[0]
    print(getattr(mem, "PercentCommittedBytesInUse"))
    sleep(1)
```
And this one will yield the same value, the value may be either taken from nowhere, or be real but stale:
```
import wmi
from time import sleep
c = wmi.WMI(".")
mem = c.Win32_PerfRawData_PerfOS_Memory()[0]
while True:
    print(getattr(mem, "PercentCommittedBytesInUse"))
    sleep(1)
```
Long story short, eval for wrappers' properties (or getattr'ing them) is not enough. One should not store instantiated counters in cobjs dictionary and recreate needed counters before each value read.

**Comment by rpratt-eng1** [comment]
I thought that something like that might be happening so I reset all the variables at the end of each loop in my while loop. But that didn't seem to work. I also tried to create a new pyperfmon object at the start of each loop, which also didn't appear to work. essentially, I did the below.  Are you saying recreate the pyperfmon object? 

 while x <= secs:
        pm = pyperfmon.pyperfmon()
        getSomeCounter and print it
        del(pm)
        x+=1

**Comment by k5md** [comment]
>Are you saying recreate the pyperfmon object?

Recreating the pyperfmon object will take much time, since it will go through all formatted counters collecting their display names.
However, with this version of pyperfmon this is the only way without monkey-patching getCounter method. To be more precise, the error is in fact that counter creation that happens on lines 117 and 121 of the pyperfmon module:
```
costr = "c.{}".format(self.connections[alias]["objects"][object]["ObjectName"])
cobjs = eval(costr)()
```
has to be done every time one tries to get desired performance counter value. This doesn't happen because of [this check](https://github.com/damies13/pyperfmon/blob/00f28ef1b6a5da8e501708c4e2d6c0b166123eff/pyperfmon/pyperfmon.py#L116
)
```
if "cobjs" not in self.connections[alias]["objects"][object]:
```
I've fixed this issue together with others in my fork, you can try it.

**Comment by damies13** [comment]
> @k5md  I've fixed this issue together with others in my fork, you can try it.

if you've already fixed it can you send a pull request? It would be good to have this fixed for everyone.

**Comment by rpratt-eng1** [comment]
Unclear as to where to get the fix. I don't see any updated files here.

**Comment by damies13** [comment]
> Unclear as to where to get the fix. I don't see any updated files here.

I'm just doing a little testing of the PR before I release it. 

If you're in a rush you can get it from https://test.pypi.org/project/pyperfmon/ when I do the release it will be version 0.2.

**Comment by damies13** [comment]
I'm not entirely sure this PR actually fixes the issue, here is the code I was using for testing, it's a variation of the old example.py


```.py
from pyperfmon import pyperfmon
import time
import subprocess
import sys

print("Start						", time.time())

res = subprocess.check_output([sys.executable, '-m', 'pip', 'show', 'pyperfmon'])
resarr = res.splitlines()
print("pyperfmon", resarr[1].decode())

print("Init Class					", time.time())

pm = pyperfmon.pyperfmon()

print("Do Loop					", time.time())

for i in range(20):
	print("\nSystem\Processor Queue Length", pm.getCounter(r"System\Processor Queue Length"))

	print(pm.getCounter(r"Processor\0\% Processor Time"))
	print(pm.getCounter(r"Processor\1\% Processor Time"))
	print(pm.getCounter(r"Processor\2\% Processor Time"))
	print(pm.getCounter(r"Processor\3\% Processor Time"))
	print(pm.getCounter(r"Processor\_Total\% Processor Time"))

	print(pm.getCounter(r"Memory\Page Writes/sec"))
	print(pm.getCounter(r"Memory\Page Reads/sec"))
	print(pm.getCounter(r"Memory\Pages/sec"))
	print(pm.getCounter(r"Memory\Cache Bytes"))
	print(pm.getCounter(r"Memory\% Committed Bytes In Use"))
	print(pm.getCounter(r"Memory\Available Bytes"))

	print("Local Counters					", time.time())

	time.sleep(10)
```

I ran https://browserbench.org/Speedometer2.0/ in a browser window in the background to vary the cpu usage, not sure of a better way to test it.

As long as nothing breaks in https://github.com/damies13/robotframework-perfmon I'll release as is. The multi-instance counters fixe makes it worthwhile.

FYI the only reason I created pyperfmon was for robotframework-perfmon, I never expected anyone else to use pyperfmon, but I'm glad you've both found it useful.

**Comment by damies13** [comment]
Updated version v0.2 now available

```
pip install -U pyperfmon
```

**Comment by k5md** [comment]
Hello! Thank you for merging my PR! I immediately tried to check if there are any issues with code since you mentioned that something might be wrong. It turns out that when installing new version (0.2) of pyperfmon, 
pyperfmon-0.2.tar.gz actually contains updated code, while pyperfmon-0.2-py3-none-any.whl contains code from version 0.1. I have no idea, why, but you can check it here https://pypi.org/project/pyperfmon/0.2/#files by downloading pyperfmon-0.2-py3-none-any.whl, extracting it and viewing pyperfmon.py there.
So when doing pip install pyperfmon==0.2 the version reported is correct, but code taken from pyperfmon-0.2-py3-none-any.whl is from old version.
>FYI the only reason I created pyperfmon was for robotframework-perfmon, I never expected anyone else to use pyperfmon, but I'm glad you've both found it useful.

That's what opensource is for, thank you for this package!

**Comment by damies13** [comment]
v0.2.1 should fix it, I think I encountered that before on another project, I think I needed to simply clear the build directory.

**Comment by damies13** [comment]
Ahhhh the evil spaces

**Comment by damies13** [comment]
Yay, it's fixed v0.2.2.1

```
System\Processor Queue Length ('System\\Processor Queue Length', 0)
('Processor\\0\\% Processor Time', '58')
('Processor\\1\\% Processor Time', '68')
('Processor\\2\\% Processor Time', '64')
('Processor\\3\\% Processor Time', '70')
('Processor\\_Total\\% Processor Time', '40')
('Memory\\Page Writes/sec', 0)
('Memory\\Page Reads/sec', 0)
('Memory\\Pages/sec', 0)
('Memory\\Cache Bytes', '68841472')
('Memory\\% Committed Bytes In Use', 50)
('Memory\\Available Bytes', '1540694016')
Local Counters                                   1677253690.7160697

System\Processor Queue Length ('System\\Processor Queue Length', 0)
('Processor\\0\\% Processor Time', '45')
('Processor\\1\\% Processor Time', '34')
('Processor\\2\\% Processor Time', '37')
('Processor\\3\\% Processor Time', '70')
('Processor\\_Total\\% Processor Time', '42')
('Memory\\Page Writes/sec', 0)
('Memory\\Page Reads/sec', 0)
('Memory\\Pages/sec', 0)
('Memory\\Cache Bytes', '67497984')
('Memory\\% Committed Bytes In Use', 50)
('Memory\\Available Bytes', '1523384320')
Local Counters                                   1677253705.1102664

System\Processor Queue Length ('System\\Processor Queue Length', 0)
('Processor\\0\\% Processor Time', '39')
('Processor\\1\\% Processor Time', '69')
('Processor\\2\\% Processor Time', '17')
('Processor\\3\\% Processor Time', '33')
('Processor\\_Total\\% Processor Time', '43')
('Memory\\Page Writes/sec', 0)
('Memory\\Page Reads/sec', 0)
('Memory\\Pages/sec', 0)
('Memory\\Cache Bytes', '68026368')
('Memory\\% Committed Bytes In Use', 52)
('Memory\\Available Bytes', '1481154560')
Local Counters                                   1677253719.5742545

System\Processor Queue Length ('System\\Processor Queue Length', 0)
('Processor\\0\\% Processor Time', '0')
('Processor\\1\\% Processor Time', '6')
('Processor\\2\\% Processor Time', '0')
('Processor\\3\\% Processor Time', '0')
('Processor\\_Total\\% Processor Time', '0')
('Memory\\Page Writes/sec', 0)
('Memory\\Page Reads/sec', 0)
('Memory\\Pages/sec', 0)
('Memory\\Cache Bytes', '67563520')
('Memory\\% Committed Bytes In Use', 52)
('Memory\\Available Bytes', '1472946176')
Local Counters                                   1677253733.7073376
```

**Comment by damies13** [comment]
> That's what opensource is for, thank you for this package!

👍  I agree, 

I would have used an existing package in robotframework-perfmon if I could have found one, but as I couldn't I thought it best to create pyperfmon and then use it for robotframework-perfmon rather than only creating robotframework-perfmon. Now it seems that was the right call.

**Comment by rpratt-eng1** [comment]
Thanks for fixing this. I was looking for an easy way to implement performance monitoring for a Benchmark called SQL Server benchmark (SSB) I wrote. It's a SQL server implementation similar to  SLOB for Oracle.

**Comment by damies13** [comment]
@rpratt-eng1 Sounds like a cool project, hope you'll be supporting multiple SQL servers, then people can use your tool to compare them.
Great to know this helped in a small way

**Comment by rpratt-eng1** [comment]
Its a command line tool that can be used to drive multiple SQL servers. It's purpose is to drive a workload to storage and measure storage performance. HammerDB doesn't do that. I do a lot of SQL perf testing and HammerDB just wasn't cutting it.  With SSB you can drive a workload with SQL server, using the Database Engine, similar to HammerDB, but, unlike HammerDB, you can drive a variety of workloads. 0-100% reads and see what the performance curve looks like for various read/write ratios.  Its not like sqlio or diskpd or fio because those are all synthetic workloads. SSB allows us to compare Oracle and other DB's performance on the same storage. It'll work with local storage or SAN and NAS as well, though it was meant to test SAN/NAS.  

https://github.com/NetApp/SQL_Storage_Benchmark
